src/hb-kbts.cc

```diff
@@ -44,59 +44,54 @@
 hb_kbts_face_data_t *
 _hb_kbts_shaper_face_data_create (hb_face_t *face)
 {
-  kbts_font *kb_font = nullptr;
-  void *data = nullptr;
-  void *scratch = nullptr;
-  void *memory = nullptr;
-  size_t scratch_size = 0;
-  size_t memory_size = 0;
-
   hb_blob_t *blob = hb_face_reference_blob (face);
 
   unsigned int blob_length;
   const char *blob_data = hb_blob_get_data (blob, &blob_length);
   if (unlikely (!blob_length))
   {
     DEBUG_MSG (KBTS, blob, "Empty blob");
-    goto fail;
+    hb_blob_destroy (blob);
+    return nullptr;
   }
 
-  data = hb_malloc (blob_length);
+  void *data = hb_malloc (blob_length);
+  if (likely (data))
+    hb_memcpy (data, blob_data, blob_length);
+
+  hb_blob_destroy (blob);
+  blob = nullptr;
+
   if (unlikely (!data))
-    goto fail;
+  {
+    DEBUG_MSG (KBTS, face, "Failed to allocate memory for font data");
+    return nullptr;
+  }
 
-  kb_font = (kbts_font *) hb_calloc (1, sizeof (kbts_font));
+  kbts_font *kb_font = (kbts_font *) hb_calloc (1, sizeof (kbts_font));
   if (unlikely (!kb_font))
   {
     hb_free (data);
-    goto fail;
+    return nullptr;
   }
 
-  // Copy the blob data because kbts_ReadFontData() will modify it.
-  hb_memcpy (data, blob_data, blob_length);
-
-  // The `data` and `memory` pointers will be freed by kbts_FreeFont().
-
-  scratch_size = kbts_ReadFontHeader (kb_font, data, blob_length);
-  scratch = hb_malloc (scratch_size);
-  memory_size = kbts_ReadFontData (kb_font, scratch, scratch_size);
-
-  memory = scratch;
-  if (memory_size > scratch_size)
-    memory = hb_realloc (memory, memory_size);
-
-  kbts_PostReadFontInitialize (kb_font, memory, memory_size);
-
-  if (unlikely (!kbts_FontIsValid(kb_font)))
+  size_t memory_size;
   {
-    DEBUG_MSG (KBTS, face, "kbts_font creation failed");
-    _hb_kbts_shaper_face_data_destroy ((hb_kbts_face_data_t *) kb_font);
-    kb_font = nullptr;
-    goto fail;
+    unsigned scratch_size = kbts_ReadFontHeader (kb_font, data, blob_length);
+    void *scratch = hb_malloc (scratch_size);
+    memory_size = kbts_ReadFontData (kb_font, scratch, scratch_size);
+    hb_free (scratch);
```

Comment:
The idea was to avoid another allocation if possible, so `scratch` was re-used if it is of enough size. (that is what `kbts_FontFromFile()` does as well).

Reply:
I saw that. But for the font I was testing, memory_size was half of scratch_size... Free then malloc of a smaller size is definitely fast. We don't need to hold onto too much memory.

Reply:
Makes sense.